fast_llm/engine/multi_stage/fast_llm_model.py

```diff
@@ -83,28 +85,83 @@ def distributed(self):
+            converter_class.save_config(checkpoint_config.path, exported_config)
+            self._save_state_dict(
+                checkpoint_config,
+                converter_class(self._base_model_config),
+                {
+                    "fast_llm_metadata": metadata,
+                    "model_config": exported_config,
+                    "format": "pt",
+                },
+            )
+        elif checkpoint_config.format == CheckpointFormat.state_dict:
+            self._save_state_dict(checkpoint_config, TrivialConverter(), metadata)
+        elif checkpoint_config.format == CheckpointFormat.distributed:
+            if self._distributed_config.rank == 0:
+                yaml.safe_dump(metadata, (checkpoint_config.path / "metadata.yaml").open("w"))
+            safetensors.torch.save_file(
+                tensors={"state_shard": self._state_shard[:num_shards]},
+                filename=checkpoint_config.path / f"rank_{self._distributed_config.rank}.safetensors",
+                metadata=_export_safetensors_metadata(metadata),
+            )
         else:
-            raise NotImplementedError(checkpoint_config.checkpoint_type)
+            raise NotImplementedError(checkpoint_config.format)
+
+    def _save_state_dict(self, checkpoint_config: CheckpointSaveConfig, converter: ModelConverter, metadata: dict):
+        with StateDictSaver(
+            checkpoint_config,
+            distributed=self._distributed,
+            metadata=metadata,
+            base_file_name=converter.base_file_name,
+        ) as context:
+            # The tensor mapping may not be one-to-one. `convert_state_dict` pops all tensors from
+            #   `fast_llm_state_dict` that are ready for conversion,
+            #   and return a dict containing the converted tensors(s).
+            #   If converting a tensor requires another one that is not yet available (e.g. for concatenation),
+            #   it will remain in `fast_llm_state_dict` until that tensor is available.
+            fast_llm_state_dict = {}
+            for i, shard_name in enumerate(
+                self._state_shard_names if checkpoint_config.optimizer_state else self._state_shard_names[:1]
+            ):
+                shard_split = self._state_shard[i].split(self._stage_shard_sizes, 0)
+                for stage, shard in zip(self._stages_on_device.values(), shard_split):
+                    for name, tensor in stage._export_shard(shard, dtype=checkpoint_config.data_type):  # noqa
+                        assert name not in fast_llm_state_dict
+                        fast_llm_state_dict[(name, shard_name)] = tensor
+                        for exported_name, exported_tensor in converter.convert_state_dict(
+                            fast_llm_state_dict, True
+                        ).items():
+                            context.add_tensor(exported_name, exported_tensor)
```

Comment:
so "external" is just state-dict saving with ad-hoc conversion?

in that case, do we need "external" as its own format?

right now:

1. `distributed`: fast internal checkpoint format, only for short-term use (i.e. within the same run and with the same (minor?) version of Fast-LLM)
2. `state_dict`: long-term internal checkpoint format, safe to use with different versions of Fast-LLM
3. `external`: like `state_dict`, but potentially lossy because optimizer states cannot be saved and tensors may be converted in a numerically lossy way.

did I capture that correctly?

Reply:
Something like that, I'll be clearing things up in the next PR. The plan is to generalize things a bit, something like
```
- AbstractSaver
    - DistributedSaver [hard-coded]
    - AbstractStateDictSaver
        - StateDictSaver [hard-coded]
        - HuggingfaceSaver
            - [model/user-specific]
        - [model/user-specific]
    - [model/user-specific]

Reply:
As for the optimizer state, the only reason it's prevented is I haven't implemented it for huggingface, it would probably not be that hard.